cpp/src/arrow/compute/exec/tpch_node.cc

```diff
@@ -0,0 +1,3483 @@
+  static constexpr int32_t kRegionKey[kRowCount] = {0, 1, 2, 3, 4};
+  static constexpr const char* kRegionNames[kRowCount] = {"AFRICA", "AMERICA", "ASIA",
+                                                          "EUROPE", "MIDDLE EAST"};
+
+  struct REGION {
+    enum {
+      R_REGIONKEY,
+      R_NAME,
+      R_COMMENT,
+      kNumColumns,
+    };
+  };
+
+  const std::unordered_map<std::string, int> kNameMap = {
+      {"R_REGIONKEY", REGION::R_REGIONKEY},
+      {"R_NAME", REGION::R_NAME},
+      {"R_COMMENT", REGION::R_COMMENT},
+  };
+
+  const std::vector<std::shared_ptr<DataType>> kTypes = {
+      int32(),
+      fixed_size_binary(kNameByteWidth),
+      utf8(),
+  };
+
+  std::shared_ptr<Schema> schema_;
+  std::vector<int> column_indices_;
+};
+
+constexpr int32_t RegionGenerator::kRegionKey[RegionGenerator::kRowCount];
+constexpr const char* RegionGenerator::kRegionNames[RegionGenerator::kRowCount];
+
+class TpchNode : public ExecNode {
+ public:
+  TpchNode(ExecPlan* plan, const char* name,
+           std::unique_ptr<TpchTableGenerator> generator)
+      : ExecNode(plan, {}, {}, generator->schema(), /*num_outputs=*/1),
+        name_(name),
+        generator_(std::move(generator)) {}
+
+  const char* kind_name() const override { return name_; }
+
+  [[noreturn]] static void NoInputs() {
+    Unreachable("TPC-H node should never have any inputs");
+  }
+
+  [[noreturn]] void InputReceived(ExecNode*, ExecBatch) override { NoInputs(); }
+
+  [[noreturn]] void ErrorReceived(ExecNode*, Status) override { NoInputs(); }
+
+  [[noreturn]] void InputFinished(ExecNode*, int) override { NoInputs(); }
+
+  Status StartProducing() override {
+    return generator_->StartProducing(
+        thread_indexer_.Capacity(),
+        [this](ExecBatch batch) { this->OutputBatchCallback(std::move(batch)); },
+        [this](int64_t num_batches) { this->FinishedCallback(num_batches); },
+        [this](std::function<Status(size_t)> func) -> Status {
+          return this->ScheduleTaskCallback(std::move(func));
+        });
+  }
+
+  void PauseProducing(ExecNode* output) override {}
+  void ResumeProducing(ExecNode* output) override {}
+
+  void StopProducing(ExecNode* output) override {
+    DCHECK_EQ(output, outputs_[0]);
+    StopProducing();
+  }
+
+  void StopProducing() override {
+    if (generator_->Abort()) std::ignore = task_group_.End();
+  }
+
+  Future<> finished() override { return task_group_.OnFinished(); }
+
+ private:
+  void OutputBatchCallback(ExecBatch batch) {
+    outputs_[0]->InputReceived(this, std::move(batch));
+  }
+
+  void FinishedCallback(int64_t total_num_batches) {
+    outputs_[0]->InputFinished(this, static_cast<int>(total_num_batches));
+    std::ignore = task_group_.End();
+  }
+
+  Status ScheduleTaskCallback(std::function<Status(size_t)> func) {
+    auto executor = plan_->exec_context()->executor();
+    if (executor) {
+      RETURN_NOT_OK(task_group_.AddTaskIfNotEnded([&] {
```

Comment:
Can you add a comment here explaining why it's ok to ignore tasks.  Maybe something like:

// The first part of a generation task is figuring out which rows to emit.
// When we reach the end of a table generator we may end up creating
// some extra final tasks.  The first one of these will mark the task group
// ended and the remaining generation tasks can be safely ignored.

Reply:
Added.